felix/nftables/ipsets.go

```diff
@@ -0,0 +1,697 @@
+			// In the dataplane, we should have updated its members above.
+			continue
+		}
+		if _, ok := s.setNameToAllMetadata[name]; !ok {
+			// Defensive: this set is not in the dataplane, and it's not
+			// one we are tracking, clean up its member tracker.
+			log.WithField("name", name).Warn("Cleaning up leaked(?) set member tracker.")
+			delete(s.mainSetNameToMembers, name)
+			continue
+		}
+		// We're tracking this set, but we didn't find it in the dataplane;
+		// reset the members set to empty.
+		members.Dataplane().DeleteAll()
+	}
+
+	return nil
+}
+
+func LegalizeSetName(setName string) string {
+	return strings.Replace(setName, ":", "-", -1)
+}
+
+func (s *IPSets) NFTablesSet(name string) *knftables.Set {
+	metadata, ok := s.setNameToAllMetadata[name]
+	if !ok {
+		return nil
+	}
+
+	flags := make([]knftables.SetFlag, 0, 1)
+	switch metadata.Type {
+	case ipsets.NFTSetTypeAddrPort:
+		// IP and port sets don't support the interval flag.
+	case ipsets.NFTSetTypeAddr:
+		// IP addr sets don't use the interval flag.
+	case ipsets.NFTSetTypeNet:
+		// Net sets require the interval flag.
+		flags = append(flags, knftables.IntervalFlag)
+	default:
+		log.WithField("type", metadata.Type).Panic("Unexpected IP set type")
+	}
+
+	return &knftables.Set{
+		Name:  LegalizeSetName(name),
+		Type:  metadata.Type.SetType(s.IPVersionConfig.Family.Version()),
+		Flags: flags,
+	}
+}
+
+// tryUpdates attempts to create and/or update IP sets.  It attempts to do the updates as a single
+// 'ipset restore' session in order to minimise process forking overhead.  Note: unlike
+// 'iptables-restore', 'ipset restore' is not atomic, updates are applied individually.
+// This function updates the set of programmed IPs - that is the IPs that were added or replaced in the IPSets
+// included by the ipsetFilter.
+func (s *IPSets) tryUpdates() error {
+	var dirtyIPSets []string
+
+	s.ipSetsWithDirtyMembers.Iter(func(setName string) error {
+		if _, ok := s.setNameToProgrammedMetadata.Desired().Get(setName); !ok {
+			// Skip deletions and IP sets that aren't needed due to the filter.
+			return nil
+		}
+		dirtyIPSets = append(dirtyIPSets, setName)
+		return nil
+	})
+
+	s.setNameToProgrammedMetadata.PendingUpdates().Iter(func(setName string, v ipsets.IPSetMetadata) deltatracker.IterAction {
+		if !s.ipSetsWithDirtyMembers.Contains(setName) {
+			dirtyIPSets = append(dirtyIPSets, setName)
+		}
+		return deltatracker.IterActionNoOp
+	})
+	if len(dirtyIPSets) == 0 {
+		s.logCxt.Debug("No dirty IP sets.")
+		return nil
+	}
+
+	start := time.Now()
+
+	// Create a new transaction to update the IP sets.
+	tx := s.nft.NewTransaction()
+
+	if s.setNameToProgrammedMetadata.Dataplane().Len() == 0 {
+		// Use the total number of IP sets that we believe we have programmed as a proxy for whether
+		// or not the table exists. If this is the first time we've programmed IP sets, make sure we
+		// create the table as part of the transaction as well.
+		tx.Add(&knftables.Table{})
```

Comment:
> Presumably this will fail if it does already exist but it was somehow empty?

no, `tx.Add` is a no-op if the object already exists

Reply:
Yeah, this will succeed even if the table exists. But I wanted this function to do nothing if nothing needed to be done.

I suppose I could instead always add the table, and then only execute the transaction if NumOperations is >1 instead of >0.